pkg/helmreconciler/rendering.go

```diff
@@ -17,117 +17,164 @@ package helmreconciler
-	"k8s.io/helm/pkg/tiller"
-	"k8s.io/helm/pkg/timeconv"
-	"k8s.io/kubernetes/pkg/kubectl"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func (h *HelmReconciler) renderCharts(input RenderingInput) (ChartManifestsMap, error) {
-	rawVals, err := yaml.Marshal(input.GetValues())
-	if err != nil {
-		return ChartManifestsMap{}, err
+func (h *HelmReconciler) renderCharts(in RenderingInput) (ChartManifestsMap, error) {
+	icp, ok := in.GetInputConfig().(*v1alpha2.IstioControlPlaneSpec)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type %T in renderCharts", in.GetInputConfig())
+	}
+	if err := validate.CheckIstioControlPlaneSpec(icp, false); err != nil {
+		return nil, err
 	}
-	config := &chart.Config{Raw: string(rawVals), Values: map[string]*chart.Value{}}
 
-	c, err := chartutil.Load(input.GetChartPath())
+	mergedICPS, err := mergeICPSWithProfile(icp)
 	if err != nil {
-		return ChartManifestsMap{}, err
-	}
-
-	renderOpts := renderutil.Options{
-		ReleaseOptions: chartutil.ReleaseOptions{
-			// XXX: hard code or use icp.GetName()
-			Name:      "istio",
-			IsInstall: true,
-			IsUpgrade: false,
-			Time:      timeconv.Now(),
-			Namespace: input.GetTargetNamespace(),
-		},
-		// XXX: hard-code or look this up somehow?
-		KubeVersion: fmt.Sprintf("%s.%s", chartutil.DefaultKubeVersion.Major, chartutil.DefaultKubeVersion.Minor),
-	}
-	renderedTemplates, err := renderutil.Render(c, config, renderOpts)
+		return nil, err
+	}
+
+	t, err := translate.NewTranslator(binversion.OperatorBinaryVersion.MinorVersion)
 	if err != nil {
-		return ChartManifestsMap{}, err
+		return nil, err
+	}
+
+	cp := controlplane.NewIstioControlPlane(mergedICPS, t)
+	if err := cp.Run(); err != nil {
+		return nil, fmt.Errorf("failed to create Istio control plane with spec: \n%v\nerror: %s", mergedICPS, err)
 	}
 
-	return collectManifestsByChart(manifest.SplitManifests(renderedTemplates)), err
+	manifests, errs := cp.RenderManifest()
+	if errs != nil {
+		err = errs.ToError()
+	}
+
+	return toChartManifestsMap(manifests), err
 }
 
-// collectManifestsByChart returns a map of chart->[]manifest.  names for subcharts
-// will be of the form <root-name>/charts/<subchart-name>, e.g. istio/charts/galley
-func collectManifestsByChart(manifests []manifest.Manifest) ChartManifestsMap {
-	manifestsByChart := make(ChartManifestsMap)
-	for _, chartManifest := range manifests {
-		pathSegments := strings.Split(chartManifest.Name, "/")
-		chartName := pathSegments[0]
-		// paths always start with the root chart name and always have a template
-		// name, so we should be safe not to check length
-		if pathSegments[1] == "charts" {
-			// subcharts will have names like <root-name>/charts/<subchart-name>/...
-			chartName = strings.Join(pathSegments[:3], "/")
+func mergeICPSWithProfile(icp *v1alpha2.IstioControlPlaneSpec) (*v1alpha2.IstioControlPlaneSpec, error) {
+	profile := icp.Profile
+
+	// This contains the IstioControlPlane CR.
+	baseCRYAML, err := helm.ReadProfileYAML(profile)
```

Comment:
would no profile specified in istio control plane spec means default profile?

Reply:
yes.